README.rst
===========
Sundowner
===========

Sundowner is a Python3 Flask application that provides an API to return JSON
objects containing Javascript timestamps of civil dawn, civil dusk, sunrise,
and sunset.

It also provides a simple web app that uses this API.

Sundowner uses `PyEphem <http://rhodesmill.org/pyephem/>`_ to calculate
astronomical parameters.

Sunrise and sunset times are calculated according to the United States Naval
Observatory specifications, whereby for dawn and dusk, atmospheric refraction is estimated by setting
observer pressure to zero and the horizon to 34 arcminutes below the normal
horizon.

Civil dawn and dusk are when the sun is 6º below the horizon.

Sundowner is licensed under the MIT License.

Dependencies
============

Sundowner has the following Python3 dependencies:

* Flask 0.10
* ephem 3.7
* itsdangerous 0.24
* Jinja2 2.7.2
* MarkupSafe 0.23
* Werkzeug 0.9.4
* python-dateutil 2.2

URL endpoints
=============

Sundowner exposes the following endpoint:

* ``/api/sun``

Parameters
----------

* ``<lat>`` and ``<lon>`` are simple decimals of latitude and longitude respectively,
  e.g. Rothera's location is expressed as -67.57 and -68.13
* ``<start>`` and ``<end>`` are start and end dates for the desired period, in ISO 8601
  format (as produced by Javascript's ``.toJSON()`` method)
* ``<output_type>`` is an optional parameter for the resulting datetimes,
  defaulting to ``iso`` for ISO 8601 format. Alternatively, use ``js``
  for Javascript timestamp, or ``dt`` for a human-readable datetime.

Output
======

The API returns a JSON object, as follows::

    {
      "days": [
        {
          "date": "2014-05-23",
          "events": {
            "civil_dawn": "2014-05-23T12:58:31.819853+00:00",
            "civil_dusk": "2014-05-23T19:59:14.923174+00:00",
            "sunrise": "2014-05-23T14:31:56.000752+00:00",
            "sunset": "2014-05-23T18:25:53.468033+00:00"
          }
        }
      ],
      "latlon": {
        "lat": "-67.57",
        "lon": "-68.13"
      }
    }

Acknowledgements
================

Sundowner uses the following components:

* Bootstrap v3.1.1 (http://getbootstrap.com)
  Licensed under the MIT License. (c) 2011-2014 Twitter, Inc
* HTML5 Shiv v3.7.0 (https://github.com/aFarkas/html5shiv)
  Licensed under the MIT License.
* jQuery v1.11.0 (http://jquery.org)
  Licensed under the MIT License. (c) 2005, 2014 jQuery Foundation, Inc.
* moment.js v2.6.0 (http://momentjs.com)
  Licensed under the MIT License. (c) 2011-2014 Tim Wood, Iskren Chernev,
  and Moment.js contributors
* Respond.js v1.4.2 (https://github.com/scottjehl/Respond)
  Licensed under the MIT License. (c) 2013 Scott Jehl
* bootstrap-datepicker v1.3.0 (http://eternicode.github.io/bootstrap-datepicker)
  Licensed under the Apache 2.0 License.

Full-text copies of the above licenses may be found at:

* http://opensource.org/licenses/MIT
* http://opensource.org/licenses/Apache-2.0


CHANGES.txt
v0.1, 2014-05-23 -- Initial release.
v0.1.1, 2014-06-10 -- CSS bugfix.
v0.1.2, 2014-06-10 -- Another CSS bugfix.
v0.1.3, 2014-06-15 -- Python code cleanup.


run_tests.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import sys, os
sys.path.insert(0, os.path.dirname(__file__))
import sundowner.test
sundowner.test.main()


sundowner/test/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Sundowner Tests

Unit tests for Sundowner
"""
import sundowner
import unittest
try:
    import simplejson as json
except ImportError:
    import json
import urllib.parse

class SundownerTestCase(unittest.TestCase):

    def setUp(self):
        """Setup for testing"""
        sundowner.app.config['TESTING'] = True
        self.app = sundowner.app.test_client()

    def tearDown(self):
        pass

    def test_index(self):
        """Check the index page loads"""
        rv = self.app.get('/')
        assert '200 OK' in rv.status
        assert b'Sundowner' in rv.data

    # Test request data
    test_request_data = {
        'lat': -67.57,
        'lon': -68.13,
        'output_type': 'iso',
        'start': '2014-06-10T03:00:00.000Z',
        'end': '2014-06-10T03:00:00.000Z'
    }
    enc_request_data = urllib.parse.urlencode(test_request_data)

    def test_api_json(self):
        """Check we get valid JSON from the api"""
        rv = self.app.get('/api/sun/?{0}'.format(self.enc_request_data))
        try:
            jdata = json.loads(rv.data.decode('utf-8'))
        except ValueError:
            raise Exception('Invalid JSON response')
        events = jdata['days'][0]['events']
        assert '200 OK' in rv.status
        assert jdata['days'][0]['date'] == '2014-06-10'
        assert events['civil_dawn'] == '2014-06-10T13:35:54.968129+00:00'
        assert events['civil_dusk'] == '2014-06-10T19:27:46.806923+00:00'
        assert events['sunrise'] == '2014-06-10T15:57:56.853457+00:00'
        assert events['sunset'] == '2014-06-10T17:05:45.835109+00:00'

if __name__ == '__main__':
    unittest.main()


sundowner/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
sundowner.py

Webservice to provide sunrise and sunset information.
All times are UTC. It is up to the client to convert to local time.
"""
import ephem
from datetime import date, datetime, tzinfo, timezone, timedelta
import dateutil.parser
from flask import Flask, jsonify, abort, render_template, url_for, request

app = Flask(__name__)

@app.route("/")
def index():
    return render_template('index.html')

@app.route('/api/sun/')
def sun():
    # Populate variables from request
    lat = request.args.get('lat', None, type=str)
    lon = request.args.get('lon', None, type=str)
    start_day = request.args.get('start', None, type=str)
    end_day = request.args.get('end', None, type=str)
    output_type = request.args.get('output_type', 'iso', type=str)

    # Default to today if no start date specified
    if start_day is None:
        start = date.today()
    else:
        start = dateutil.parser.parse(start_day).date()

    if end_day is None:
        end = start
    else:
        end = dateutil.parser.parse(end_day).date()

    days_range = (end + timedelta(days=1) - start).days
    days = [start + timedelta(days=i) for i in range(days_range)]
    times = [{'date': day.isoformat(), 'events': list_times(lat, lon, day, output_type)} for day in days]
    output = {'latlon': {'lat': lat, 'lon': lon}, 'days': times}
    return jsonify(output)

def parse_date(day):
    """
    Parses a date in YYYYMMDD format and returns a Date object.
    """
    try:
        # Wrap input in int() in case of preceeding zeros
        output = date(int(day[:4]), int(day[4:6]), int(day[6:8]))
    except ValueError or TypeError:
        abort(500)
    return output

def list_times(latitude, longitude, day, output_type):
    """
    Uses calculate_sunrise() to work out the relevent events for a given day.

    Args:
      day (date): Date object
      latitude (str): Latitude
      longitude (str): Longitude
      output_type (str): 'js' = Javascript timestamp in ms,
        'dt' = Python datetime object, 'iso' = ISO 8601 format

    Returns:
      events (dict): dictionary containing events as Unix timestamps in ms

    Raises:
      TypeError: If day, latitude, longitude, or elevation are invalid.
    """
    location = ephem.Observer()
    location.lat = latitude
    location.lon = longitude
    location.pressure = 0
    # -6º is civil twilight; –0:34º with 0 pressure is the NOAA spec for
    # sunrise/set to take into account atmospheric refraction
    events = {'civil_dawn': ('-6', True),
              'sunrise': ('-0:34', True),
              'sunset': ('-0:34', False),
              'civil_dusk': ('-6', False)}
    output = {event: calculate_sunrise(location, day, events[event][0],
                events[event][1], output_type) for event in events}
    return output

def calculate_sunrise(location, day, horizon, dawn='True', output_type='iso'):
    """
    Calculate sunrise/set time and return it as a Julian Date.

    Args:
      location (object): an ephem.Observer object
      day (float): the Julian Day or Date
      horizon (str): the angle of the desired horizon
        e.g. astronomical dawn is -18
      dawn (bool): Dawn (True) or dusk (False)
      output_type (str): 'js' = Javascript timestamp in ms,
        'dt' = Python datetime object, 'iso' = ISO 8601 format

    Returns:
      suntime (int): naive (UTC) timestamp integer of the sunrise/sunset

    Raises:
      CircumpolarError: If sun does not rise or set during the Julian Day.
    """
    location.date = day
    location.horizon = horizon
    sun = ephem.Sun()
    sun.compute(location)
    if dawn is True:
        try:
            suntime = location.next_rising(sun).datetime()
        except ephem.CircumpolarError:
            suntime = None
    else:
        try:
            suntime = location.next_setting(sun).datetime()
        except ephem.CircumpolarError:
            suntime = None
    if suntime is not None:
        suntime = suntime.replace(tzinfo=timezone.utc)
        if output_type == 'js':
            # Parse datetime object as a Javascript datetime
            # (milliseconds since Unix epoch)
            output_time = round(suntime.timestamp() * 1000)
        elif output_type == 'dt':
            # Just a simple Python datetime object
            output_time = suntime
        elif output_type == 'iso':
            # A datetime in ISO 8601 format
            output_time = suntime.isoformat()
    else:
        output_time = suntime
    return output_time

if __name__ == "__main__":
    app.run()
